### ISSUE TITLE
Unable to send messages since 3.1.0.20 update

### ISSUE THREAD

**Issue by quimic** [signalapp/Signal-iOS#4347]
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
The app gets stuck trying to send a message. This happens for each and every message since the update.
To see an example please look at the debug log (link below):
- At 09:55:08:661 I tried to send a message. I kept the app in the foreground, waiting for about 10mn.
- At 10:05:35 I clicked on the home button
- At 10:05:37 I tapped on the Signal icon which reopened the app on the lock screen. I never got a chance to unlock (I waited about 2 and a half minutes)
- At 10:08:15 I forced quit the app, then started it again. The message finally got sent

### Steps to reproduce
- Open Signal
- Type new message
- Tap the send icon

**Actual result:**
The "sending" animation shows (dotted line circle), and it stays like that until app is restarted.

**Expected result:**
Message should appear as sent after a few seconds.

### Device info

**Device**: iPhone 6S (iPhone8,1)

**iOS version**: 13.3 (17C54)

**Signal version:** 3.1.0 (3_1_0_20)

### Link to debug log
https://debuglogs.org/4fbaa5c13ca8330080c0683bc84967a7a1bf513a6c310563741a54d0995732f3.zip

**Comment by richlv** [comment]
Seeing the same problem with the latest version. Additional problems that might be related - if not, I'd be happy to create new issues.

* App often crashes.
* Messages are sometimes lost - this might happen when force-closing the app fairly soon after attempting to send a message.
* It is not possible to send images (choosing an image shows a spinner).

**Comment by michaelkirk-signal** [comment]
Hi @richlv - can you include a crash and debug log? 

https://github.com/signalapp/Signal-iOS/wiki/Submitting-crash-reports

**Comment by richlv** [comment]
Debuglog at https://debuglogs.org/dbbdc41fea16f95afc151fb4e509b41f6c2ef3ce6520a943294ec4d61e58b3a2.zip .

Not seeing any recent Signal ips files.

**Comment by quimic** [comment]
> Hi @richlv - can you include a crash and debug log?
> 
> https://github.com/signalapp/Signal-iOS/wiki/Submitting-crash-reports

I just sent a few crash reports.

**Comment by richlv** [comment]
Possibly related - phone reports that over the past 24 hours Signal was responsible for 52% of the consumed power, and shows "Background Activity" next to it.

**Comment by richlv** [comment]
Usual pattern seems to be:
* After a longer period of time, send one message to a recipient.
* Try to send another - takes a terribly long time, app might crash.

**Comment by the-x-at** [comment]
Same problem here. 

- Signal v 3.1.0.20
- iOS 12.4.2
-  iPhone 5s. 

Longest time to send a single message was 1.5 hours (90 minutes), on average 10 minutes to send a single message. Frequent crashes. Sometimes a message to be sent is lost after crash/force quitting [home button double press] the app, especially during the first minute of trying to send.
Most messages are received quickly, though. Sending, however is virtually impossible. For me, Signal Beta is currently unusable.

**Comment by Ginari** [comment]
Same issue here!
- Signal 3.1.0.20
- iOS 13.3.
- iPhone 7

Also it seems the iPhone gets quite hot when sending. Not entirely sure if related though. Quick fix would be much appreciated!

**Comment by the-x-at** [comment]
To make things even worse, I cannot paste messages into the text box. So messages that were in the process of being sent followed by an app crash and message loss cannot be copied back. I have to retype each message. This might be totally unrelated, so I will monitor the behavior and file a separate issue if needed.

**Comment by Nerd2k** [comment]
I can confirm all of the aforementioned issues and behaviours!

Device: iPad mini 4
iPadOS version: 13.3
Signal version: 3.1.0.20

**Comment by gprison** [comment]
I can confirm all of the aforementioned issues and behaviours!

Device: iPhone 6S
iOS version: 13.2.3
Signal version: 3.1.0.20

**Comment by michaelkirk-signal** [comment]
Hi all - If you are experiencing this issue, please attach a debug log to this issue right after it happens, or you can email it to me directly at michael@signal.org.

Please include, as quimic helpfully did, the timestamp of events, like when you tried to send, and when the message finally sent. What I'm seeing in the logs is mostly a bunch of successful sends, so it's hard to track down the problematic section without some additional context from you.

Also, take a look for crash reports as described in:
https://github.com/signalapp/Signal-iOS/wiki/Submitting-crash-reports

> I just sent a few crash reports.

Where did you send them @quimic?  

If you emailed them to support, could you make sure you referenced this GH issue? Or if you mean that you uploaded them via the Test Flight crash interface, could you also either attach them here or email them to support@signal.org and reference this GH issue? Otherwise I don't really have a way to separate them from other unrelated crash logs uploaded to Test Flight.

Thanks for any help tracking this down!

**Comment by quimic** [comment]
> Hi all - If you are experiencing this issue, please attach a debug log to this issue right after it happens, or you can email it to me directly at [michael@signal.org](mailto:michael@signal.org).
> 
> Please include, as quimic helpfully did, the timestamp of events, like when you tried to send, and when the message finally sent. What I'm seeing in the logs is mostly a bunch of successful sends, so it's hard to track down the problematic section without some additional context from you.
> 
> Also, take a look for crash reports as described in:
> https://github.com/signalapp/Signal-iOS/wiki/Submitting-crash-reports
> 
> > I just sent a few crash reports.
> 
> Where did you send them @quimic?
> 
> If you emailed them to support, could you make sure you referenced this GH issue? Or if you mean that you uploaded them via the Test Flight crash interface, could you also either attach them here or email them to [support@signal.org](mailto:support@signal.org) and reference this GH issue? Otherwise I don't really have a way to separate them from other unrelated crash logs uploaded to Test Flight.
> 
> Thanks for any help tracking this down!

Hi Michael,
I did send the crash reports to support@signal.org with the full GH issue title (and #) on the 12/29 (at 9:29 UTC+1). I just forwarded the message to you (with the files obviously).
Let me know if you need more data.
M.

**Comment by michaelkirk-signal** [comment]
Some questions for those experiencing this issue...

1. Did you notice this issue when running 3.0?

2. Approximately how long were you running 3.0 before running 3.1? (like not at all, or a couple days?)

3. Do you have screen protection enabled?

4. Do you have screen lock enabled?

5. What duration is your screen lock set to?

6. Anything else you can tell us that might be "uncommon" about your setup? e.g. have you denied contact access? Are you on a VPN?

And if you haven't already, please send a debug log with the time that you sent a message that hung.

**Comment by quimic** [comment]
> 1. Did you notice this issue when running 3.0?

No
> 2. Approximately how long were you running 3.0 before running 3.1? (like not at all, or a couple days?)

 I always update Signal very quickly (within 24h of the update), so I ran it for as long as it was available before 3.1.
> 3. Do you have screen protection enabled?

Yes
> 4. Do you have screen lock enabled?

Yes
> 5. What duration is your screen lock set to?

Instant
> 6. Anything else you can tell us that might be "uncommon" about your setup? e.g. have you denied contact access? Are you on a VPN?

I have indeed denied contact access. But no other uncommon setting I can think of.

> And if you haven't already, please send a debug log with the time that you sent a message that hung.

You can find the original debug log file link in my issue text.  If you look at the specific actions and precise timing I specified in the text, you should be able to see when Signal was trying to send the message.

Meanwhile, I ran a test of sending a message without screen lock. I got mostly the same behaviour (the spinning circle). While I was writing this email, my phone's screen went black (normal behaviour), so I unlocked the phone and opened Signal again, finding the same situation (I was not blocked by the lock screen, but the message was still marked as sent).
I waited some more, and when I went back to Signal, the message had been sent this time.
You can find the corresponding debug log here: https://debuglogs.org/d7b06b131d498d5d1631460dd8e71c37ff44218ea9afc54bab4497141a8f0534.zip

I don't think the lock screen is what causes the issue (and my latests test seems to confirm that). It just makes the situation more complicated, because while Signal is busy sending the message, I cannot get in, and I eventually have to force quit the app to get back in.

Hope this helps understanding what is going on…

**Comment by quimic** [comment]
FYI I have Signal 3.0.6.0 installed on an iPad (I did not update Signal after I noticed the issues on my phone with the v3.1). It has the exact same settings (screen lock, screen protection, no access to contacts), and I don't have any issue sending messages…

**Comment by gprison** [comment]
Some questions for those experiencing this issue...

    Did you notice this issue when running 3.0?
No.

    Approximately how long were you running 3.0 before running 3.1? (like not at all, or a couple days?)
longer. until the Appstore said there is a new update.
    Do you have screen protection enabled?
Yes.

    Do you have screen lock enabled?
Yes.

    What duration is your screen lock set to?
1 minute

    Anything else you can tell us that might be "uncommon" about your setup? e.g. have you denied contact access? Are you on a VPN?
No.

**Comment by Nerd2k** [comment]
>     1. Did you notice this issue when running 3.0?
no
> 
>     2. Approximately how long were you running 3.0 before running 3.1? (like not at all, or a couple days?)
I check for updates in the app store almost every day. So I have updated to 3.1 essentially like 1 day after it became available.
> 
>     3. Do you have screen protection enabled?
yes
> 
>     4. Do you have screen lock enabled?
no
> 
>     5. What duration is your screen lock set to?
not applicable
> 
>     6. Anything else you can tell us that might be "uncommon" about your setup? e.g. have you denied contact access? Are you on a VPN?
no denied contact access, no VPN, nothing special really

**Comment by michaelkirk-signal** [comment]
Thanks so much for the info @quimic, @gprison, and @Nerd2k.

If any of you would be willing to run a diagnostic build, could you follow up by emailing michael@signal.org with your device UDID?

To get your device UDID:

- attach your iPhone to your mac
- Launch iTunes, and select the device management button (1) then toggle the UDID display by clicking "serial number" (2)

![image](https://user-images.githubusercontent.com/36971200/71383703-63521c80-2592-11ea-90b7-3d3f3e3109f3.png)

- Once iTunes is displaying the device UDID, copy the text of the UDID into an email to michael@signal.org
<img width="1212" alt="Screen Shot 2019-12-23 at 2 35 38 PM" src="https://user-images.githubusercontent.com/36971200/71383545-bc6d8080-2591-11ea-90ac-6b6872c7bc53.png">

I'll follow up with the diagnostic build as soon as it's ready.

**Comment by michaelkirk-signal** [comment]
I have a theory for what's happening here - the theory involves idling for a bit in a very long conversation. 

The long conversation doesn't necessarily need to be the conversation in which you're failing to send a message, just that you've opened the long conversation since launching the app. This could kick off a bunch of local work which would bog down the database, preventing other database access, like that which is required for message sending.

If someone experiencing the issue could send me their device UDID I can get you a build to see if that resolves your issue.

**Comment by the-x-at** [comment]
Did you notice this issue when running 3.0?
No, never before. Have joined the testing program in August 2018.

    Approximately how long were you running 3.0 before running 3.1? (like not at all, or a couple days?)
I update to the newest Signal beta version as soon as it gets out (i.e. I get notified by Testflight), this is within hours.

    Do you have screen protection enabled?
Yes

    Do you have screen lock enabled?
Yes

    What duration is your screen lock set to?
30 seconds

    Anything else you can tell us that might be "uncommon" about your setup? e.g. have you denied contact access? Are you on a VPN?
I am always in power saving mode. No VPN. Bug happens on WiFi and on mobile service provider connections.

There are quite a few partners where I can send messages almost instantly. However, one partner's communication channel is constantly "congested". From this person I receive messages very quickly and almost without delay, sending is taking between 2 minutes and 40 minutes (measured over roughly 24 hours communication). This person has an iPhone with the non-testing up to date Signal version 3.1. It's also the person I have a very long, possibly the longest chat history.

The following debug log is from a text that has been sent after 17 minutes on Dec 22, 2019. Pushed the send button at 19:02 CET (Vienna/Berlin/Rome tz), screen got locked after my usual 30 seconds, and after 17 minutes I looked at the Signal screen, the app crashed, after a restart the message had been sent. The log has been generated immediately after the crash. 

https://debuglogs.org/42387fa0d0b9b4fd3032417000f1c49099a3513a32c097d69bfd796d475d66a8.zip

**Comment by Nerd2k** [comment]
> I have a theory for what's happening here - the theory involves idling for a bit in a very long conversation.
> 
> The long conversation doesn't necessarily need to be the conversation in which you're failing to send a message, just that you've opened the long conversation since launching the app. This could kick off a bunch of local work which would bog down the database, preventing other database access, like that which is required for message sending.
> 
> If someone experiencing the issue could send me their device UDID I can get you a build to see if that resolves your issue.

Thank you very much @michaelkirk-signal. That theory seems very plausible to me - for me the issue also occurs within/is triggered by a very long conversation, so I can confirm that.

Unfortunately, I'm not a test pilot participant. Please, could anyone else check out that diagnostic build? Thank you!

**Comment by michaelkirk-signal** [comment]
You don’t need to be a test flight user to install an Ad Hoc build. 

You will need to send me your device UDID as described in the post.

**Comment by the-x-at** [comment]
In support of @michaelkirk-signal's hypothesis, I virtually don't see any issues with Signal when not touching the "killer-thread" (the one with the very high number of messages stored) after a fresh start. This even includes the aforementioned issue by @richlv concerning sending images and my issue with pasting text into the edit field. This works as long as I don't open that killer-thread and send something there. 
The irony of this bug is that you cannot properly communicate with those you communicate most. :-D

**Comment by Nerd2k** [comment]
> The irony of this bug is that you cannot properly communicate with those you communicate most. :-D

Yeah, absolutely 😄

**Comment by Nerd2k** [comment]
Any news regarding the bug + possible fix?

**Comment by charlesmchen-signal** [comment]
@Nerd2k We're going to have a fix out as soon as the app store reopens from the holiday.  Could be Sunday.

**Comment by richlv** [comment]
A possible workaround for images - sending them from the "Pictures" app, sharing option.
If Signal is closed, this method works much better. In an emergency, it could even be used to send a simple message, attached to a small picture :)

**Comment by michaelkirk-signal** [comment]
> @Nerd2k We're going to have a fix out as soon as the app store reopens from the holiday. Could be Sunday.

The 3.1.1 release is now available in the app store. Please update and let us know if you continue to experience this issue.

**Comment by richlv** [comment]
Messages can again be sent, thank you :)
Adding images seems to be slowish, but I'm not sure whether that's any slower than before, or just my imagination by now.

**Comment by Nerd2k** [comment]
> Messages can again be sent, thank you :)

Confirmed. Thank you @charlesmchen-signal and @michaelkirk-signal 👍

> Adding images seems to be slowish, but I'm not sure whether that's any slower than before, or just my imagination by now.

Well, I have the feeling that this has been the case since V3 in general.
